src/highdicom/seg/sop.py

```diff
@@ -61,6 +75,723 @@
+            segment. If False (the default), segments are binary and stacked
+            down the last dimension of the output array.
+        relabel: bool, optional
+            If True and ``combine_segments`` is ``True``, the output segment
+            numbers are relabelled into the range ``0`` to
+            ``len(segment_numbers)`` (inclusive) accoring to the position of
+            the original segment numbers in ``segment_numbers`` parameter.  If
+            ``combine_segments`` is ``False``, this has no effect.
+
+        Yields
+        ------
+        Iterator[Tuple[int, int, int]]:
+            Iterator of indices required to construct the requested mask. Each
+            triplet denotes the (output frame index, segmentation frame index,
+            output segment number) representing a list of "instructions" to
+            create the requested output array by copying frames from the
+            segmentation dataset.
+
+        """
+        # Run query to create the iterable of indices needed to construct the
+        # desired pixel array. The approach here is to create two temporary
+        # tables in the SQLite database, one for the desired source UIDs, and
+        # another for the desired segments, then use table joins to arrive with
+        # the referenced UIDs table and the frame LUT to arrive at the relevant
+        # rows, before clearing up the temporary tables.
+
+        # Create temporary table of desired frame numbers
+        table_name = 'TemporarySOPInstanceUIDs'
+        column_defs = [
+            'OutputFrameIndex INTEGER UNIQUE NOT NULL',
+            'SourceSOPInstanceUID VARCHAR UNIQUE NOT NULL'
+        ]
+        column_data = enumerate(source_sop_instance_uids)
+
+        # Construct the query The ORDER BY is not logically necessary
+        # but seems to improve performance of the downstream numpy
+        # operations, presumably as it is more cache efficient
+        query = (
+            'SELECT '
+            '    T.OutputFrameIndex,'
+            '    L.FrameNumber - 1,'
+            '    S.OutputSegmentNumber '
+            'FROM TemporarySOPInstanceUIDs T '
+            'INNER JOIN InstanceUIDs U'
+            '    ON T.SourceSOPInstanceUID = U.SOPInstanceUID '
+            'INNER JOIN FrameLUT L'
+            '    ON U.SOPInstanceUID = L.ReferencedSOPInstanceUID '
+            'INNER JOIN TemporarySegmentNumbers S'
+            '    ON L.SegmentNumber = S.SegmentNumber '
+            'ORDER BY T.OutputFrameIndex'
+        )
+
+        with self.temp_table(
+            table_name=table_name,
+            column_defs=column_defs,
+            column_data=column_data,
+        ):
+            with self.temp_segment_table(
+                segment_numbers=segment_numbers,
+                combine_segments=combine_segments,
+                relabel=relabel
+            ):
+                yield self._db_con.execute(query)
+
+    @contextmanager
+    def get_indices_by_source_frame(
```

Comment:
```suggestion
    def iter_indices_by_source_frame(
```

Reply:
It may be helpful to describe somewhere why the two different methods are needed. I am assuming that the `iter_indices_by_source_instance()` is needed if a multi-frame segmentation image references multiple single-frame image instances and the `iter_indices_by_source_frame()` is needed if a multi-frame segmentation image references multiple frames of a single multi-frame image instance. Is that understanding correct?